veps/sig-storage/incremental-backup.md

```diff
@@ -0,0 +1,374 @@
+
+**Handling Checkpoints Across VM Restarts**
+One challenge in Kubevirt is that Libvirt is re-created each time a VM restarts, causing all existing libvirt checkpoints to be lost. To enable incremental backups after a restart—rather than falling back to a full backup, checkpoints metadata must be provided to libvirt during VM initialization. This allows libvirt to redefine the checkpoints, ensuring that an incremental backup can be continued as expected.
+To address this, the VM controller will use the stored metadata in the VirtualMachineBackupChain CRs during startup. It will iterate through all relevant chains and re-create the checkpoints in Libvirt based on the stored metadata. This ensures that incremental backups can resume seamlessly, even after a VM restart.
+
+### Offline backups
+
+An initial proposed solution involves starting the VM in a paused state to perform the backup, and then either unpausing or shutting it down once the backup is complete. However, this approach introduces additional considerations and trade-offs, which justify a separate, dedicated discussion. As such, this section is intentionally left without a formal API definition for now and only online backup will be supported at first.
+
+### State interruptions during backup
+
+In the case of an online backup if the user tries to stop the VM there will be a built-in [solution](https://issues.redhat.com/browse/RHEL-8067) that Libvirt is currently working on. Until then, we may need to handle it on the Kubevirt side by preventing changes to the VM state during the backup or before stopping the VM we would wait for the VirtualMachineBackup job to complete and only then allow the VM to stop.
+
+### VM crash
+
+In the event of a VM crash, the CBT information may become corrupted and must be discarded. QEMU provides an API to verify if the dirty bitmaps are valid. Upon VM restart, while redefining the checkpoints, we will check the validity of the bitmaps. If a corrupted bitmap is detected, the checkpoints will be discarded and deleted from the `VirtualMachineBackupChain` indicating that next backup will be a full backup.
+
+### Migration and Backup
+
+Migration and backup operations are mutually exclusive and cannot be performed simultaneously.
+During migration, the bitmaps of all previous backups are automatically transferred along with the disk image. The only additional step required is similar to what is done on VM restarts: the checkpoints must be redefined for the newly initialized Libvirt instance.
+TODO: Confirm whether the VM state PVC also needs to be migrated.
+If migration of the state PVC is needed but not possible, all existing checkpoints will be lost, and the next backup will have to fall back to a full backup.
```

Comment:
Can we be more clear about what happens when a VM is migrated during backup? I assume the backup is cancelled/aborted? Even if migration can theoretically work while backing up, user may not be backing up to RWX PVC. Given that Filesystem PVC is required for backup I expect that to be the most common case

Reply:
I assume that migration has a bigger priority? if so I guess backup will be aborted.

> Even if migration can theoretically work while backing up, user may not be backing up to RWX PVC. Given that Filesystem PVC is required for backup I expect that to be the most common case
Not sure I understand the question here. From what I understad from @pipo we dont really care if the storage and VM state pvc is RWX since we can just create a new one in the migration target, the data will be kept in the raw image and the bitmaps will be streamed in the qemu migration